libbeat/processors/actions/copy_fields_test.go

```diff
@@ -0,0 +1,116 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package actions
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+)
+
+func TestCopyFields(t *testing.T) {
+
+	var tests = map[string]struct {
+		FromTo   fromTo
+		Input    common.MapStr
+		Expected common.MapStr
+	}{
+		"copy string from message to message_copied": {
+			FromTo: fromTo{
+				From: "message",
+				To:   "message_copied",
+			},
+			Input: common.MapStr{
+				"message": "please copy this line",
+			},
+			Expected: common.MapStr{
+				"message":        "please copy this line",
+				"message_copied": "please copy this line",
+			},
+		},
+		"copy string from nested key nested.message to top level field message_copied": {
+			FromTo: fromTo{
+				From: "nested.message",
+				To:   "message_copied",
+			},
+			Input: common.MapStr{
+				"nested": common.MapStr{
+					"message": "please copy this line",
+				},
+			},
+			Expected: common.MapStr{
+				"nested": common.MapStr{
+					"message": "please copy this line",
+				},
+				"message_copied": "please copy this line",
+			},
+		},
+		"copy string from fieldname with dot to message_copied": {
+			FromTo: fromTo{
+				From: "dotted.message",
+				To:   "message_copied",
+			},
+			Input: common.MapStr{
+				"dotted.message": "please copy this line",
+			},
+			Expected: common.MapStr{
+				"dotted.message": "please copy this line",
+				"message_copied": "please copy this line",
+			},
+		},
+		"copy number from fieldname with dot to dotted message.copied": {
+			FromTo: fromTo{
+				From: "message.original",
```

Comment:
What happens if you copy `message.original` to `message`? Will it fail?

Reply:
I set `message.original` as a single key containing a dot, so it does not lead to an error when someone wants to manipulate the submaps under `message`. It is a different key.

If the test input was different e.g. the key `message` was already present

```golang
Input: common.MapStr{
    "message": common.MapStr{
        "original": 42,
    },
}
```

It would fail and tell the user to drop the field `message` first, before attempting to copy stuff.

This is a more realistic test case, so I am adding it also.

Reply:
Interesting, good to know. Could you add a test case for both and one of them would fail. The reason I'm asking also for the failing one is because in case we change the behaviour of the above in the future, we will know that it has an impact.

Reply:
The test with the description `"copy number from hierarchical message.original to top level message which fails"` will fail if the behaviour changes, as the new field would show up in the output. I have also added a new test case which demonstrates what happens when the key contains a dot and is present as a top level key in the fields. If `MapStr` is changed one of the tests will fail.